firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/TimestampSerializer.kt

```diff
@@ -4,63 +4,105 @@ import dev.gitlive.firebase.FirebaseDecoder
 import dev.gitlive.firebase.FirebaseEncoder
 import kotlinx.serialization.KSerializer
 import kotlinx.serialization.SerializationException
+import kotlinx.serialization.descriptors.buildClassSerialDescriptor
 import kotlinx.serialization.descriptors.element
 import kotlinx.serialization.encoding.Decoder
 import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.encoding.decodeStructure
+import kotlinx.serialization.encoding.encodeStructure
 
-sealed class AbstractTimestampSerializer<T>(
-    private val isNullable: Boolean
-) : KSerializer<T> {
 
-    override val descriptor = buildClassSerialDescriptor("Timestamp", isNullable = isNullable) {
-        isNullable = this@AbstractTimestampSerializer.isNullable
+/** A serializer for [BaseTimestamp]. If used with [FirebaseEncoder] performs serialization using native Firebase mechanisms. */
+object BaseTimestampSerializer : KSerializer<BaseTimestamp> {
+    override val descriptor = buildClassSerialDescriptor("Timestamp") {
         element<Long>("seconds")
         element<Int>("nanoseconds")
+        element<Boolean>("isServerTimestamp")
     }
 
-    protected fun encode(encoder: Encoder, value: FirebaseTimestamp?) {
+    override fun serialize(encoder: Encoder, value: BaseTimestamp) {
         if (encoder is FirebaseEncoder) {
-            encoder.value = value?.let {
-                when(value) {
-                    is FirebaseTimestamp.Value -> value.value
-                    is FirebaseTimestamp.ServerValue -> FieldValue.serverTimestamp()
-                    is FirebaseTimestamp.ServerDelete -> FieldValue.delete
-                }
+            // special case if encoding. Firestore encodes and decodes Timestamp without use of serializers
+            encoder.value = when (value) {
+                Timestamp.ServerTimestamp -> FieldValue.serverTimestamp().platformValue
+                is Timestamp -> value.platformValue
+                else -> throw SerializationException("Cannot serialize $value")
             }
         } else {
-            throw IllegalArgumentException("This serializer must be used with FirebaseEncoder")
+            encoder.encodeStructure(descriptor) {
+                when (value) {
+                    Timestamp.ServerTimestamp -> {
+                        encodeLongElement(descriptor, 0, 0)
+                        encodeIntElement(descriptor, 1, 0)
+                        encodeBooleanElement(descriptor, 2, true)
+                    }
+                    is Timestamp -> {
+                        encodeLongElement(descriptor, 0, value.seconds)
+                        encodeIntElement(descriptor, 1, value.nanoseconds)
+                        encodeBooleanElement(descriptor, 2, true)
```

Comment:
If we still use this flag, here should be `false` then

Reply:
Yeah, this should be false

Reply:
fixed